### PR TITLE
Introduce an implementation that calculates the crc without using a lookup table

### PR DESCRIPTION
**Pull request by KillingSpark**
This PR introduces an implementation for the Crc width 32, that calculates the crc without using a lookup table.

It also adds an explicit choice for the Bytewise implementation, which is the current default behaviour for `Crc<u32>` 

Additionally it reorganizes the modules a little more neatly for the crc32 instead of having `crc32, crc32_fast, ...` it's now:
```
src
├── crc32
    ├── bytewise.rs
    ├── default.rs
    ├── nolookup.rs
    └── slice16.rs
```

This will allow easily and (invisibly to users) changing the default implementation.

This should also be used as the template for providing `Nolookup<>` and `Bytewise<>` for the other widths.

### Benches
```
crc32/nolookup          time:   [137.87 µs 137.98 µs 138.10 µs]
                        thrpt:  [113.14 MiB/s 113.24 MiB/s 113.33 MiB/s]

crc32/bytewise          time:   [29.316 µs 29.330 µs 29.345 µs]
                        thrpt:  [532.46 MiB/s 532.74 MiB/s 532.98 MiB/s]

crc32/slice16           time:   [4.5155 µs 4.5248 µs 4.5324 µs]
                        thrpt:  [3.3666 GiB/s 3.3723 GiB/s 3.3792 GiB/s]+
```

### CODE REVIEW

**Comment by akhilles** [comment]
Thanks for submitting the PR! I'm a bit swamped this week, but I'll try to review this over the next couple days.

**Comment by KillingSpark** [comment]
Don't worry if it takes longer. I know how it can be sometimes :)

File: src/lib.rs [hunk 1]
```diff
@@ -77,3 +96,8 @@ pub struct Digest<'a, I: Implementation> {
     crc: &'a Crc<I>,
     value: I::Width,
 }
+
+// Convenient type aliases
+pub type Crc32Slice16 = Crc<Slice16<u32>>;
+pub type Crc32Bytewise = Crc<Bytewise<u32>>;
+pub type Crc32Nolookup = Crc<Nolookup<u32>>;
```
Comment (akhilles) [inline src/lib.rs:103]: I lean towards not having these aliases. In recommended usage, the full type is used infrequently I want to minimize API surface as much as possible. Users can always add their own aliases.

File: src/lib.rs [hunk 2]
```diff
@@ -67,6 +84,8 @@ impl<W: Width> Implementation for W {
     type Table = [W; 256];
 }
 
+/// Crc with pluggable implementations. `I`can be either a unsigned integer type (e.g. u32) to get the default implementation for that width
+/// or one of `Bytewise<>`, `Nolookup<>`, or `Slice16<>` to choose the used implementation
```
Comment (akhilles) [inline src/lib.rs:88]:
```suggestion
/// Crc with pluggable implementations ([Nolookup], [Bytewise], [Slice16]).
/// To choose the default implemenation, use the [Width] directly (e.g. `Crc<u32>`).
```

**Comment by akhilles** [comment]
Overall, the changes LGTM. Just have some minor nitpicks.

File: src/crc32.rs [hunk 3]
```diff
@@ -1,83 +1,200 @@
-use super::{Algorithm, Crc, Digest};
-use crate::table::crc32_table;
+mod bytewise;
```
[the remark this answers is not in the document]
Comment (KillingSpark) [reply src/crc32.rs:1]: From the rust docs:

> Note: Prior to rustc 1.30, using mod.rs files was the way to load a module with nested children. It is encouraged to use the new naming convention as it is more consistent, and avoids having many files named mod.rs within a project.

I can still do it if you want to of course :)

File: src/crc32/slice16.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        Self { algorithm, table }
+    }
+
+    pub const fn checksum(&self, bytes: &[u8]) -> u32 {
+        let mut crc = self.init(self.algorithm.init);
+        crc = self.update(crc, bytes);
+        self.finalize(crc)
+    }
+
+    const fn init(&self, initial: u32) -> u32 {
```
[the remark this answers is not in the document]
Comment (KillingSpark) [reply src/crc32/slice16.rs]: Done

File: src/lib.rs [hunk 5]
```diff
@@ -36,22 +36,39 @@ pub use crc_catalog::*;
...
 
-/// For some widths there is a faster implementation using a 16x larger lookup table. Use it with `Crc<Slice16<W>>`
+/// Implementation using a 16 * 256 entries lookup table. Use it with `Crc<Slice16<W>>`
 pub struct Slice16<W: Width>(core::marker::PhantomData<W>);
 
+/// Implementation using a 256 entries lookup table. Use it with `Crc<Bytewise<W>>`
+pub struct Bytewise<W: Width>(core::marker::PhantomData<W>);
+
+/// Implementation using no lookup table. Use it with `Crc<Nolookup<W>>`
+pub struct Nolookup<W: Width>(core::marker::PhantomData<W>);
```
[the remark this answers is not in the document]
Comment (KillingSpark) [reply src/lib.rs]: Yep I like it

File: src/lib.rs (hunk 1 above)
Comment (KillingSpark) [reply src/lib.rs]: :+1:

**Comment by KillingSpark** [comment]
Apart from the crc32.rs -> mod.rs rename this LGTM. 

`Introduce an implementation that calculates the crc32 without using a lookup table` would probably be a good squashed commit message

File: src/crc32.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (akhilles) [reply src/crc32.rs:1]: Ah didn't know about this, let's stick with the new convention.

**Comment by KillingSpark** [comment]
I'll make some more PRs for the other widths. Do you prefer the same separation for slice16/notable impls or should I do one PR per width that adds both slice16/notable at once?

**Comment by akhilles** [comment]
> I'll make some more PRs for the other widths. Do you prefer the same separation for slice16/notable impls or should I do one PR per width that adds both slice16/notable at once?

Up to you, whatever's easier.

**Pull request merged**